Reimplementing blocking awaiting for Futures so only the blockers have t...

viktorklang opened the pull request:
...o pay for it.

Changing so that only blockers pay for signalling by using CDL + onComplete, saving monitor-usage for non-blockers.

Review by: @phaller & @adriaanm

viktorklang committed:
Reimplementing much of the DefaultPromise methods
Optimizations:
1) Avoiding to call 'synchronized' in tryComplete and in tryAwait
2) Implementing blocking by using an optimized latch so no blocking ops for non-blockers
3) Reducing method size of isCompleted to be cheaper to inline
4) 'result' to use Try.get instead of patmat

adriaanm commented:
please follow the commit message conventions documented in https://github.com/scala/scala/wiki/Pull-Request-Policy

adriaanm commented on src/library/scala/concurrent/impl/Promise.scala:
```diff
@@ -59,34 +61,24 @@ private[concurrent] object Promise {
-        case Duration.Inf        => awaitUnbounded
-        case Duration.MinusInf   => isCompleted
-        case f: FiniteDuration   => if (f > Duration.Zero) awaitUnsafe(f.fromNow, f) else isCompleted
+        case e if e eq Undefined => throw new IllegalArgumentException("cannot wait for Undefined period")
```
do you need `eq`? if `==` is ok, `case Undefined` will do

viktorklang commented:
Changed commit message, this one is good to go from my side.

adriaanm commented:
@axel22, since @phaller is travelling, could you have a look please?

If this needs to go into 2.10.2, it has to be merged this week.

adriaanm commented:
PS: in principle, note that the PR policy states that the submitter of the PR is responsible for getting the PR reviewed and merged in time.

viktorklang commented:
It's not crucial for 2.10.2

As a sidenote, I don't see how feasible it is to put the responsibility of reviews on the submitter as it is outside his/her control.

adriaanm commented:
Who does have control, then? Centralizing reviewer-chasing does not scale.

som-snytt commented:
Adriaan knows a guy, who knows a guy, who can exert some persuasion.

One of those guys doesn't even know the other guy's name.

We community who are not so connected in the Scala underworld do feel the possibly-justified burden of making a proposed change worth somebody's while.  (For example, by an exchange of review favors.)  (Otherwise, we could just keep it in a personal fork and maintain it ourselves.  That is, until the Scala world realized how earth-shattering our commit was and broke ranks and adopted our fork as the scala de jure.  "De jure" is different from "du jour", right?)

axel22 commented:
After taking a look at this commit, and also:
http://www.docjar.com/html/api/java/util/concurrent/CountDownLatch.java.html
http://www.docjar.com/html/api/java/util/concurrent/locks/AbstractQueuedSynchronizer.java.html
I conclude the change is correct, as far as I can see.

On the other hand, I am not sure I understand the reason for the change. Are we fixing a bug (current incorrect behaviour) or addressing a performance issue here?

It seems to me that this change might slow down the `Await.*` calls after the future has completed:

```
      val c = new CountDownLatch(1)
      onComplete(_ => c.countDown())(InternalCallbackExecutor)
      c.await()
```

After the future is completed, calling `Await.*` will create a `CountDownLatch`, which in turn creates a `Sync` object. We then create a closure object to install an `onComplete` callback. This guy in turn creates a `CallbackRunnable` object. That's (at least) four objects being created to do a simple read.

I understand that calling `Await` methods is not a recommended usage pattern, but I'm not sure if it's necessary that we create these many objects for a simple read. Shouldn't we instead call `isCompleted` first? Otherwise, with this implementation, the premise that blocking is bad would surely be validated :)

As for the case where a future is not yet completed when an `Await.*` call happens, was the old `awaitUnbounded` implementation really slower than what `doAcquireInterruptibly` below does?

http://www.docjar.com/html/api/java/util/concurrent/locks/AbstractQueuedSynchronizer.java.html

viktorklang commented:
Hi Alex,

Good observations indeed!
The reason for the change is twofold:
1) Safety - Avoid synchronizing on "this", this can be problematic both for weird interactions with user code and accidental interactions if implementations use lazy vals etc
2) Performance - Doing it right should be cheap, and doing it wrong should cost, with this change synchronization overhead is only paid by those who want to do blocking, i.e. those who want synchronization.

As you say, we could change isCompleted first and then only after that allocate the CountDownLatch, I'll see if that's possible.

The problem with the old awaitUnbounded is that it's quite a bit of liability, at least now we know that we don't have signalling-bugs in our code and I trust Doug a lot more than I do myself.  As for AQS vs CTD you have a point, it will be cheaper to go with AQS than CTD due to the indirection overhead of the internal Sync of the CTD, I'll see if I can get that to play nice as it would keep allocations down.

Thanks for reviewing Alex!

viktorklang commented:
Oh, Alex, btw, isCompleted is called before tryAwait in def ready:

```
 def ready(atMost: Duration)(implicit permit: CanAwait): this.type =
  if (isCompleted || tryAwait(atMost)) this
  else throw new TimeoutException("Futures timed out after [" + atMost + "]")
```

ready is then used for result as:

```
def result(atMost: Duration)(implicit permit: CanAwait): T =
  ready(atMost).value.get match {
    case Failure(e)  => throw e
    case Success(r) => r
  }
```

Thinking about it we can simplify result and remove the patmat and just use Try.get, I'll do that as well.

axel22 commented:
Ah, true - I missed that part - `isCompleted` is really used before creating synchronization objects.
I see that we eliminate the `notifyAll` in the common case which is supposed to be more expensive - this makes sense.

About performance, I cannot help feeling that we could have avoided the countdown latch by creating a dummy object when someone calls `Await.*` and calling `notifyAll` from the callback:

```
class Dummy { var completed = false }
val a = new Dummy
onComplete(_ => a.synchronized { a.completed = true; a.notifyAll() })(InternalCallbackExecutor)
a.synchronized { while (!a.completed) a.wait() }
```

Seems to me like this does less work, then again, we would have to have some performance numbers to argue - something that the current build system does not support.

About safety, I can see your reasoning. It would be good to have a test case that tries to capture the lazy val/`Await` interaction, though. Since none of our future implementations have lazy vals in themselves, I was thinking about something like this:

```
def test()
  var f: Future[Int] = null
  class Lazy {
    lazy val field: Int = Await.ready(f, Duration.Inf)
    lazy val anotherField: Int = 0
  }
  val x = new Lazy
  f = future { x.anotherField }
  x.field
}
```

However, it seems to me that the change will not fix the situation, since `x.field` in the main thread will obtain the monitor of `x` and not release it when the `LockSupport.park` gets called. Assuming that the future arrives later, it will not be able to initialize `anotherField`, since it will not be able to obtain the `x` monitor.

That said, I would like to see a concrete test case, but cannot think of one.

Other than that, the change LGTM.

viktorklang commented:
Hi Alex,

thanks for the comments, take a look at the latest code here,
I created a CompletionLatch based on AQS and implemented Function1, this eliminates 2 allocations:
1) the callback function
2) the indirection between CTD and its inner Sync

I also moved the isCompleted check into tryAwait for containment.

The CompletionLatch is based on Doug's example from jsr166 which is pubic domain and there's virtually nothing to it, which makes it easier to maintain.

Thoughts/comments?

axel22 commented:
Hi!

Yes, this one does look much better to me.

From the licensing point of view, I am not sure if we are allowed to/want to have the comment about where the completion latch comes from, but maybe that should be reviewed by someone with more knowledge about licenses than me :)

Also, I think that some of your previous checks (in `isCompleted`) were ok - last time I checked, the `_: <class>` pattern produces `instanceof` checks at the bytecode level, so the two should be equivalent.

axel22 commented on src/library/scala/concurrent/impl/Promise.scala:
```diff
@@ -52,70 +55,69 @@ private[concurrent] object Promise {
-      case _: Try[_] => true
-      case _ => false
-    }
+    override def isCompleted: Boolean = getState.isInstanceOf[Try[_]]
```
Should be equivalent - I think both produce `instanceof` in bytecode.

viktorklang commented:
Alex, since we are including jsr166 almost verbatim in the scala-library I think it should be more than fine. It is public domain after all, and it's attributed (even though the example itself isn't rocket science, some attribution is in order imo)

The instanceOf check clocks in at 1/4 of the former code tho :)

axel22 commented:
LGTM

adriaanm commented:
please include some of the great reasoning about your optimizations in the code and the commit message
(also, commit title is too long)

PR comments are great for reviewing, but they don't last

viktorklang commented:
Commit message updated

adriaanm commented:
Thanks!
